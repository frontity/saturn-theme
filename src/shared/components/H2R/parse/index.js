import { parse } from 'himalaya';
import he from 'he';
import htmlMap from './htmlMap';
import svgMap from './svgMap';

// Map of lowercased HTML and SVG attributes to get their camelCase version.
const attrMap = { ...htmlMap, ...svgMap };

// Adapts the Himalaya AST Specification v1
// (see https://github.com/andrejewski/himalaya/blob/v1.0.1/text/ast-spec-v1.md)
// to our format, with the following structure:
// {
//   type: 'element',
//   component: 'p',
//   props: {},
//   children: [],
//   parent: Element,
// }
const adaptNode = (element, parent) => {
  const { type, tagName, attributes, children } = element;

  // Do not adapt 'text' or 'comment' nodes,
  // just decode escaped html characters.
  if (type !== 'element') {
    const content = he.decode(element.content);
    // Return null if empty (or only whitespaces).
    return content.trim().length ? { type, content, parent } : null;
  }

  const adapted = {
    type,
    component: tagName,
    props: attributes.reduce((attrs, { key, value }) => {
      if (key === 'class') {
        attrs.className = value;
      } else if (!/^on/.test(key)) {
        const camelCaseKey = attrMap[key.toLowerCase()];
        attrs[camelCaseKey || key] = value;
      }
      return attrs;
    }, {}),
  };

  // Parent cannot be changed.
  Object.defineProperty(adapted, 'parent', { value: parent });

  adapted.children = children.reduce((all, child) => {
    const childAdapted = adaptNode(child, adapted);
    // ignore null children
    if (childAdapted) all.push(childAdapted);
    return all;
  }, []);

  return adapted;
};

// Parse HTML code using Himalaya's parse function first, and then
// adapting each node to our format.
export default html =>
  parse(html).reduce((htmlTree, element) => {
    const adapted = adaptNode(element, null);
    if (adapted) htmlTree.push(adapted);
    return htmlTree;
  }, []);
