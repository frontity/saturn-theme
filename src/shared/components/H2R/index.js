/* eslint-disable jest/no-disabled-tests, no-console */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { inject } from 'mobx-react';
import { compose } from 'recompose';
import { withTheme } from 'styled-components';
import { compact } from 'lodash';
import parse from './parse';

class H2R extends Component {
  static propTypes = {
    html: PropTypes.string.isRequired,
    processors: PropTypes.arrayOf(PropTypes.shape({})),
    extraProps: PropTypes.shape({}),
    stores: PropTypes.shape({}).isRequired,
    theme: PropTypes.shape({}).isRequired,
  };

  static defaultProps = {
    processors: [],
    extraProps: {},
  };

  constructor(props) {
    super(props);
    this.process = this.process.bind(this);
    this.handleNode = this.handleNode.bind(this);
    this.handleNodes = this.handleNodes.bind(this);
  }

  process(node) {
    const { processors, extraProps, stores, theme } = this.props;
    const { htmlTree } = this;

    const payload = { extraProps, stores, theme, htmlTree };

    for (let i = 0; i < processors.length; i += 1) {
      const proc = processors[i];

      // Test processor function
      let isMatch = false;
      try {
        isMatch = proc.test(node, payload);
      } catch (e) {
        // ignore error
      }

      // Apply processor function
      if (isMatch) {
        try {
          const processed = proc.process(node, payload);

          // Return true if was removed
          if (!processed) return true;

          // Do a shallow merge if node is different
          if (node !== processed) {
            Object.assign(node, processed);
          }
        } catch (e) {
          // Show error message and continue processing
          console.error(e);
        }
      }
    }

    // Return false if node was not removed
    return false;
  }

  handleNodes(nodes) {
    if (!nodes) return null;

    for (let i = 0; i < nodes.length; i += 1) {
      nodes[i] = this.handleNode(nodes[i], i);
    }

    const compacted = compact(nodes);
    if (compacted.length > 1) return compacted;
    if (compacted.length === 1) return compacted[0];
    return null;
  }

  handleNode(node, index) {
    const isRemoved = this.process(node);

    // Return nothing for 'comment' nodes
    if (isRemoved || node.type === 'comment') return null;

    // Return the content of 'text' nodes
    if (node.type === 'text') return node.content;

    // Convert 'element' nodes to React
    return (
      <node.component {...node.props} key={index}>
        {this.handleNodes(node.children)}
      </node.component>
    );
  }

  render() {
    // const isBrowser = typeof window !== 'undefined';

    // if (isBrowser) {
    //   window.performance.mark('parse');
    // }

    this.htmlTree = parse(this.props.html);

    // if (isBrowser) {
    //   window.performance.mark('handle');
    //   window.performance.measure('ðŸ”¥ h2r [parse]', 'parse', 'handle');
    // }

    const toReturn = this.handleNodes(this.htmlTree);

    // if (isBrowser) {
    //   window.performance.mark('end');
    //   window.performance.measure('ðŸ”¥ h2r [handle]', 'handle', 'end');
    // }

    return toReturn;
  }
}

export default compose(
  withTheme,
  inject(({ stores }) => {
    const processors = stores.theme.h2r.processorsByPriority;
    return { stores, processors };
  }),
)(H2R);
