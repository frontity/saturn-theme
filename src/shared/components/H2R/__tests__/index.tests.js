/* eslint-disable react/prop-types */
import React from 'react';
import renderer from 'react-test-renderer';
import { types } from 'mobx-state-tree';
import { Provider as MobxProvider } from 'mobx-react';
import { ThemeProvider } from 'styled-components';
import H2R from '..';

// Test cases (h2r):
// without processors
// with a processor that test returns false
// with a processor that removes the node
// with a processor that changes component
// with a processor that changes component to React
// with a processor that adds/removes attributes
// with a processor that removes children
// with a processor that adds children
// with a processor that adds a node after it
// with a processor that uses extraProps
// with two or more processors
// with two or more processors (the first one fails while processing)
// with two or more processors that compete

const html = `
<!--This is a comment-->
<p id="paragraph">
  This is an html <span>example</span> for testing purposes.
</p>
<div class="hello">
  <img src="http://example.com/img.jpg">
</div>
`;

const ThemeStoreMock = types.model({
  h2r: types.optional(types.frozen, {
    processorsByPriority: [],
  }),
});

const renderH2R = (processors, extraProps = {}) => {
  const theme = ThemeStoreMock.create({
    h2r: { processorsByPriority: processors },
  });

  return renderer
    .create(
      <MobxProvider stores={{ theme }}>
        <ThemeProvider theme={{ color: 'red' }}>
          <H2R html={html} extraProps={extraProps} />
        </ThemeProvider>
      </MobxProvider>,
    )
    .toJSON();
};

describe('H2R', () => {
  test('without processors', () => {
    expect(renderH2R([])).toMatchSnapshot();
  });

  test('with a processor that test fails', () => {
    const testFail = {
      test: () => {
        throw Error("A processor's test has failed but nothing bad happened!");
      },
      process: () => ({ type: 'text', content: 'TEST HAS FAILED' }),
    };

    expect(renderH2R([testFail])).toMatchSnapshot();
  });

  test('with a processor that removes the node', () => {
    const removeImg = {
      test: ({ component }) => component === 'img',
      process: () => null,
    };

    expect(renderH2R([removeImg])).toMatchSnapshot();
  });

  test('with a processor that changes type', () => {
    const commentToText = {
      test: ({ type }) => type === 'comment',
      process: () => ({ type: 'text' }),
    };

    expect(renderH2R([commentToText])).toMatchSnapshot();
  });

  test('with a processor that changes component', () => {
    const toBlockquote = {
      test: ({ props }) => props.id === 'paragraph',
      process: () => ({ component: 'blockquote' }),
    };

    expect(renderH2R([toBlockquote])).toMatchSnapshot();
  });

  test('with a processor that changes component to React', () => {
    const Blockquote = ({ id, children }) => (
      <div id={id}>
        This is a React component
        <blockquote>{children}</blockquote>
      </div>
    );

    const toBlockquote = {
      test: ({ props }) => props.id === 'paragraph',
      process: () => ({ component: Blockquote }),
    };

    expect(renderH2R([toBlockquote])).toMatchSnapshot();
  });

  test('with a processor that adds/removes attributes', () => {
    const idToClassName = {
      test: ({ props }) => props.id === 'paragraph',
      process: ({ props }) => ({ props: { className: props.id } }),
    };

    expect(renderH2R([idToClassName])).toMatchSnapshot();
  });

  test('with a processor that removes children', () => {
    const removeChildren = {
      test: ({ props }) =>
        props.id === 'paragraph' || props.className === 'hello',
      process: () => ({ children: null }),
    };

    expect(renderH2R([removeChildren])).toMatchSnapshot();
  });

  test('with a processor that adds a node after it', () => {
    const addNextNode = {
      test: ({ component }) => component === 'span',
      process: node => {
        const { children } = node.parent;
        const index = children.indexOf(node);
        children.splice(index + 1, 0, { type: 'text', content: ' code ' });
        return node;
      },
    };

    expect(renderH2R([addNextNode])).toMatchSnapshot();
  });

  test('with a processor that uses extraProps', () => {
    const idToClassName = {
      test: ({ props }) => props.id === 'paragraph',
      process: (_, { extraProps }) => ({ props: { id: extraProps.newId } }),
    };

    expect(renderH2R([idToClassName], { newId: 'test' })).toMatchSnapshot();
  });

  test('with two or more processors', () => {
    const commentToText = {
      test: ({ type }) => type === 'comment',
      process: () => ({ type: 'text' }),
    };

    const toBlockquote = {
      test: ({ props }) => props.id === 'paragraph',
      process: () => ({ component: 'blockquote' }),
    };

    const removeImg = {
      test: ({ component }) => component === 'img',
      process: () => null,
    };

    expect(
      renderH2R([commentToText, toBlockquote, removeImg]),
    ).toMatchSnapshot();
  });

  test('with two or more processors (the first one fails while processing)', () => {
    const processFails = {
      test: ({ props }) => props.id === 'paragraph',
      process: () => {
        throw Error('A processor has failed but nothing bad happened!');
      },
    };

    const toBlockquote = {
      test: ({ props }) => props.id === 'paragraph',
      process: () => ({ component: 'blockquote' }),
    };

    expect(renderH2R([processFails, toBlockquote])).toMatchSnapshot();
  });

  test('with two or more processors that try to change the same node', () => {
    const toFigure = {
      test: ({ props }) => props.id === 'paragraph',
      process: ({ props, ...others }) => ({
        component: 'figure',
        props: {},
        children: [
          {
            type: 'element',
            component: 'img',
            src: 'https://example.com/test.png',
          },
          { ...others },
        ],
      }),
    };

    const toBlockquote = {
      test: ({ props }) => props.id === 'paragraph',
      process: () => ({ component: 'blockquote' }),
    };

    expect(renderH2R([toFigure, toBlockquote])).toMatchSnapshot();
  });
});
