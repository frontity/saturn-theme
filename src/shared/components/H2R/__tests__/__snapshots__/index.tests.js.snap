// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`H2R with a processor that adds a node after it 1`] = `
Array [
  <p
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     code 
     for testing purposes.

  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with a processor that adds/removes attributes 1`] = `
Array [
  <p
    className="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with a processor that changes component 1`] = `
Array [
  <blockquote
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </blockquote>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with a processor that changes component to React 1`] = `
Array [
  <div
    id="paragraph"
  >
    This is a React component
    <blockquote>
      
  This is an html 
      <span>
        example
      </span>
       for testing purposes.

    </blockquote>
  </div>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with a processor that changes type 1`] = `
Array [
  "This is a comment",
  <p
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with a processor that removes children 1`] = `
Array [
  <p
    id="paragraph"
  />,
  <div
    className="hello"
  />,
]
`;

exports[`H2R with a processor that removes the node 1`] = `
Array [
  <p
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </p>,
  <div
    className="hello"
  />,
]
`;

exports[`H2R with a processor that test fails 1`] = `
Array [
  <p
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with a processor that uses extraProps 1`] = `
Array [
  <p
    id="test"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with two or more processors (the first one fails while processing) 1`] = `
Array [
  <blockquote
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </blockquote>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R with two or more processors 1`] = `
Array [
  "This is a comment",
  <blockquote
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </blockquote>,
  <div
    className="hello"
  />,
]
`;

exports[`H2R with two or more processors that try to change the same node 1`] = `
Array [
  <figure>
    <img />
    <p>
      
  This is an html 
      <span>
        example
      </span>
       for testing purposes.

    </p>
  </figure>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;

exports[`H2R without processors 1`] = `
Array [
  <p
    id="paragraph"
  >
    
  This is an html 
    <span>
      example
    </span>
     for testing purposes.

  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
]
`;
